File: src/python/core/src/dicebear/resolver.py
"""Deterministic resolution of options against a style, with memoization."""

from __future__ import annotations

from collections.abc import Callable
from typing import Any, TypeVar, cast

from .errors import CircularColorReferenceError
from .options import Options, Range
from .prng import Prng
from .style import Style
from .style.component import Component
from .style.component_variant import ComponentVariant
from .utils.color import Color as ColorUtil

T = TypeVar("T")


class Resolver:
    """Bundles the :class:`Style`, validated :class:`Options`, and a seeded
    :class:`Prng`, exposing them as memoized named accessors.

    The memo also serves as the informational snapshot returned by
    :meth:`resolved` — every value the resolver picks during one resolution
    lands there, except for the raw seed.
    """

    def __init__(self, style: Style, options: Options) -> None:
        self._style = style
        self._options = options
        self._prng = Prng(self.seed())
        self._color_resolving: list[str] = []
        self._result: dict[str, Any] = {}

    def seed(self) -> str:
        # Deliberately not memoized — the seed is the only input kept out of the
        # resolved() snapshot, so a serialized avatar never leaks it.
        return self._options.seed() or ""

    def size(self) -> int | None:
        return self._memo("size", self._options.size)

    def id_randomization(self) -> bool:
        return self._memo(
            "idRandomization", lambda: self._options.id_randomization() or False
        )

    def title(self) -> str | None:
        return self._memo("title", self._options.title)

    def flip(self) -> str:
        return self._memo(
            "flip", lambda: self._prng.pick("flip", self._options.flip()) or "none"
        )

    def font_family(self) -> str:
        return self._memo(
            "fontFamily",
            lambda: (
                self._prng.pick("fontFamily", self._options.font_family())
                or "system-ui"
            ),
        )

    def font_weight(self) -> int | float:
        return self._memo(
            "fontWeight",
            lambda: self._prng.pick("fontWeight", self._options.font_weight()) or 400,
        )

    def scale(self) -> float:
        return self._memo_float("scale", self._options.scale(), 1.0)

    def border_radius(self) -> float:
        return self._memo_float("borderRadius", self._options.border_radius(), 0.0)

    def rotate(self) -> float:
        return self._memo_float("rotate", self._options.rotate(), 0.0)

    def translate_x(self) -> float:
        return self._memo_float("translateX", self._options.translate_x(), 0.0)

    def translate_y(self) -> float:
        return self._memo_float("translateY", self._options.translate_y(), 0.0)

    def variant(self, name: str) -> str | None:
        """Select a variant for the given component.

        The pool the PRNG draws from is built from the per-component
        ``{name}Variant`` option and the global ``tags`` filter (see
        :meth:`_variant_weights`). Only variants that exist in the style
        definition are considered.
        """
        return self._memo(f"{name}Variant", lambda: self._resolve_variant(name))

    def _resolve_variant(self, name: str) -> str | None:
        component = self._style.components().get(name)

        if component is None or not self._is_visible(name, component):
            return None

        return self._prng.weighted_pick(
            f"{name}Variant", self._variant_weights(component)
        )

    def _variant_weights(self, component: Component) -> dict[str, int | float]:
        """Build the name → weight map the PRNG draws a variant from.

        The per-component ``{name}Variant`` option is more specific than the
        global ``tags`` filter, so it takes precedence: when set, it fully
        governs the component's pool (its named variants, weighted by the option)
        and the tags filter is ignored for that component. The tags filter
        applies only where the user gave no explicit ``{name}Variant`` (see
        :meth:`_tag_filtered_names`), and falls back to every variant when
        neither is set.

        Names the style does not define are dropped, and an empty
        ``{name}Variant`` (or an empty tag result) yields no variant.
        """
        variants = component.variants()
        named = self._options.component_variant(component.source_name())
        weights: dict[str, int | float] = {}

        if named is not None:
            names: list[str] = list(named.keys())
        elif len(self._options.tags()) > 0:
            names = self._tag_filtered_names(variants)
        else:
            names = list(variants.keys())

        for v in names:
            variant = variants.get(v)

            if variant is not None:
                weights[v] = named[v] if named is not None else variant.weight()

        return weights

    def _tag_filtered_names(self, variants: dict[str, ComponentVariant]) -> list[str]:
        """Narrow a component's variants to the names satisfying the global
        ``tags`` filter, applying the parsed :meth:`Options.tags` tokens in one
        pass over the pool:

        - A positive ``cat:value`` token is an axis-scoped allow. Within each
          category some allow mentions, a variant is kept only if it carries no
          tag in that category (untouched) or matches one of the allowed values
          (OR within the category). Distinct allowed categories combine with
          AND, and a category no allow mentions is left unconstrained. A bare
          positive ``cat`` token carries no value, so it imposes no constraint (a
          no-op).
        - A negative ``!cat``/``!cat:value`` token disallows, dropping every
          variant carrying any tag in ``cat`` (bare) or the exact ``cat:value``
          tag. Disallows are checked alongside allows but always win.

        Returns the surviving variant names in definition order.
        """
        allows: dict[str, list[str]] = {}
        disallows: list[tuple[str, str | None]] = []

        for token in self._options.tags():
            if token.negated:
                disallows.append((token.category, token.value))
            elif token.value is not None:
                allows.setdefault(token.category, []).append(token.value)

        # Materialize the allow groups once, not on every variant.
        allow_groups = list(allows.items())

        names: list[str] = []

        for name, variant in variants.items():
            allowed = all(
                not variant.has_tag(category)
                or any(variant.has_tag(category, value) for value in values)
                for category, values in allow_groups
            )
            disallowed = any(
                variant.has_tag(category, value) for category, value in disallows
            )

            if allowed and not disallowed:
                names.append(name)

        return names

    def color(self, name: str) -> list[str]:
        return self._memo(f"{name}Color", lambda: self._resolve_color(name))

    def color_fill(self, name: str) -> str:
        return self._memo(
            f"{name}ColorFill",
            lambda: (
                self._prng.pick(f"{name}ColorFill", self._options.color_fill(name))
                or "solid"
            ),
        )

    def color_angle(self, name: str) -> float:
        return self._memo_float(
            f"{name}ColorAngle", self._options.color_angle(name), 0.0
        )

    def component_transform(self, name: str) -> dict[str, float]:
        """Pick the rotate/translateX/translateY/scale values for a component.

        Memoized per ``name``, so the four values land in :meth:`resolved` as
        ``{name}Rotate`` / ``{name}TranslateX`` / ``{name}TranslateY`` /
        ``{name}Scale`` for downstream introspection.
        """
        component = self._style.components().get(name)

        return {
            "rotate": self._memo_float(
                f"{name}Rotate",
                component.rotate() if component is not None else None,
                0.0,
            ),
            "translateX": self._memo_float(
                f"{name}TranslateX",
                component.translate().x() if component is not None else None,
                0.0,
            ),
            "translateY": self._memo_float(
                f"{name}TranslateY",
                component.translate().y() if component is not None else None,
                0.0,
            ),
            "scale": self._memo_float(
                f"{name}Scale",
                component.scale() if component is not None else None,
                1.0,
            ),
        }

    def resolved(self) -> dict[str, Any]:
        """Return an informational snapshot of every value the resolver picked.

        Unset entries (``None``) are filtered out so they disappear on JSON
        encode, mirroring the JS behavior. Whole-number floats are emitted as
        ints (``1``, not ``1.0``) so ``to_json()`` is byte-identical to the JS
        and Rust ports. The raw seed is excluded.
        """
        return {k: _num_value(v) for k, v in self._result.items() if v is not None}

    def _probability(self, component: Component) -> int | float:
        """Return the visibility probability (0–100) for the component.

        Aliases read the source component's user-set probability so a single
        ``{source}Probability`` option propagates to every alias of the source.
        """
        raw = self._options.component_probability(component.source_name())

        return raw if raw is not None else component.probability()

    def _is_visible(self, name: str, component: Component) -> bool:
        return self._prng.bool(f"{name}Probability", self._probability(component))

    def _resolve_color(self, name: str) -> list[str]:
        """Resolve a named color to its final stop list, applying contrast
        sorting and ``notEqualTo`` filtering from the style definition.
        """
        user_colors = self._options.color(name)
        style_color = self._style.colors().get(name)

        if user_colors is not None:
            source = user_colors
        elif style_color is not None:
            source = style_color.values()
        else:
            source = []

        candidates = [ColorUtil.to_hex(c) for c in source]
        fill = self.color_fill(name)
        stops = 1 if fill == "solid" else self._color_fill_stops(name)

        if style_color is None:
            return self._prng.shuffle(f"{name}Color", candidates)[:stops]

        # Detect circular references (e.g. a.contrastTo = b, b.contrastTo = a).
        if name in self._color_resolving:
            raise CircularColorReferenceError([*self._color_resolving, name])

        self._color_resolving.append(name)
        contrast_to = style_color.contrast_to()
        not_equal_to = style_color.not_equal_to()

        try:
            if contrast_to is not None:
                ref = self.color(contrast_to)
                ref_color = ref[0] if len(ref) > 0 else None

                if ref_color is not None:
                    candidates = ColorUtil.sort_by_contrast(candidates, ref_color)

            if len(not_equal_to) > 0:
                excluded: list[str] = []

                for ref_name in not_equal_to:
                    excluded.extend(self.color(ref_name))

                candidates = ColorUtil.filter_not_equal_to(candidates, excluded)
        finally:
            self._color_resolving.pop()

        # Skip shuffle when sorted by contrast to preserve the ordering.
        ordered = (
            candidates
            if contrast_to is not None
            else self._prng.shuffle(f"{name}Color", candidates)
        )

        return ordered[:stops]

    def _color_fill_stops(self, name: str) -> int:
        range_ = self._options.color_fill_stops(name)

        return (
            2 if range_ is None else self._prng.integer(f"{name}ColorFillStops", range_)
        )

    def _memo_float(self, key: str, range_: Range | None, fallback: float) -> float:
        return self._memo(
            key, lambda: fallback if range_ is None else self._prng.float(key, range_)
        )

    def _memo(self, key: str, compute: Callable[[], T]) -> T:
        if key in self._result:
            return cast(T, self._result[key])

        value = compute()
        self._result[key] = value

        return value


def _num_value(value: Any) -> Any:
    """Encode a numeric snapshot value the way the JS/Rust ports do.

    A whole-number float becomes a JSON integer (``1``, not ``1.0``) so the
    resolved-options snapshot is byte-identical across ports. ``bool`` is an
    ``int`` subclass but never a ``float``, so it is left untouched; non-floats,
    fractional floats, and non-finite floats (``NaN`` / ``inf``) pass through.
    """
    if isinstance(value, float) and value.is_integer() and abs(value) < 2**53:
        return int(value)

    return value

File: src/go/core/internal/render/resolver.go
package render

import (
	"github.com/dicebear/dicebear-go/v10/color"
	"github.com/dicebear/dicebear-go/v10/internal/errs"
	"github.com/dicebear/dicebear-go/v10/internal/prng"
	"github.com/dicebear/dicebear-go/v10/internal/style"
)

// resolver derives every deterministic value for an avatar from the style, the
// user options, and a seeded PRNG, exposing them as memoized named accessors.
//
// Every value the resolver picks is recorded in a snapshot (resolved), which
// the public Avatar.JSON returns. The raw seed is deliberately excluded. The
// PRNG is key-based and order-independent, so the snapshot doubles as the memo
// and the only mutable state besides the colorResolving stack that detects
// circular color references.
type resolver struct {
	style          *style.Style
	options        *options
	rng            *prng.Prng
	colorResolving []string
	result         map[string]any
	order          []string
	tags           *tagFilter
}

func newResolver(s *style.Style, opts *options) *resolver {
	seed, _ := opts.seed()
	return &resolver{
		style:   s,
		options: opts,
		rng:     prng.New(seed),
		result:  map[string]any{},
	}
}

func strIdentity(s string) string { return s }

// record inserts a snapshot value if the key has not been recorded yet (the
// first write wins, mirroring the JS #memo). Insertion order is tracked so the
// JSON envelope can emit keys in resolution order, like the JS/Rust ports.
func (r *resolver) record(key string, value any) {
	if _, ok := r.result[key]; !ok {
		r.result[key] = value
		r.order = append(r.order, key)
	}
}

// seed is deliberately not recorded — it is the one input kept out of the
// resolved snapshot, so a serialized avatar never leaks it.
func (r *resolver) seed() string {
	s, _ := r.options.seed()
	return s
}

func (r *resolver) size() *float64 {
	v := r.options.size()
	if v != nil {
		r.record("size", *v)
	} else {
		r.record("size", nil)
	}
	return v
}

func (r *resolver) idRandomization() bool {
	v := false
	if p := r.options.idRandomization(); p != nil {
		v = *p
	}
	r.record("idRandomization", v)
	return v
}

// animation returns the global animation switch, recorded under `animation`,
// default false. Deliberately without PRNG involvement: whether an avatar
// animates must not depend on the seed.
func (r *resolver) animation() bool {
	v := false
	if p := r.options.animation(); p != nil {
		v = *p
	}
	r.record("animation", v)
	return v
}

// animationPlays reports whether one timeline plays. A named timeline follows
// its ${name}Animation switch when the user set one, recorded under that key,
// and the global switch otherwise. Unnamed timelines (name is "") always
// follow the global switch. The switch is read only when asked for, so a name
// the style does not carry leaves nothing in the snapshot.
func (r *resolver) animationPlays(name string) bool {
	if name == "" {
		return r.animation()
	}

	v, ok := r.options.animationFor(name)
	if !ok {
		return r.animation()
	}

	r.record(name+"Animation", v)
	return v
}

func (r *resolver) animationSpeed() float64 {
	return r.floatOpt("animationSpeed", r.options.animationSpeed(), 1)
}

// animationSpeedFor returns the speed factor of one timeline. A named
// timeline plays at its ${name}AnimationSpeed option when the user set one,
// drawn under that key, and at the global factor otherwise. Unnamed timelines
// (name is "") always follow the global factor. The option is resolved only
// when asked for, so a name the style does not carry leaves nothing in the
// snapshot.
func (r *resolver) animationSpeedFor(name string) float64 {
	if name == "" {
		return r.animationSpeed()
	}

	rng := r.options.animationSpeedFor(name)
	if rng == nil {
		return r.animationSpeed()
	}

	return r.floatOpt(name+"AnimationSpeed", rng, 1)
}

// animationDelay returns the start offset in seconds shared by every
// timeline, drawn under `animationDelay`, default 0.
func (r *resolver) animationDelay() float64 {
	return r.floatOpt("animationDelay", r.options.animationDelay(), 0)
}

// animationDelayFor returns the start offset of one timeline. A named
// timeline uses its ${name}AnimationDelay option when the user set one, drawn
// under that key, and the global offset otherwise. Unnamed timelines (name is
// "") always follow the global offset. Mirrors animationSpeedFor.
func (r *resolver) animationDelayFor(name string) float64 {
	if name == "" {
		return r.animationDelay()
	}

	rng := r.options.animationDelayFor(name)
	if rng == nil {
		return r.animationDelay()
	}

	return r.floatOpt(name+"AnimationDelay", rng, 0)
}

func (r *resolver) title() (string, bool) {
	s, ok := r.options.title()
	if ok {
		r.record("title", s)
	} else {
		r.record("title", nil)
	}
	return s, ok
}

func (r *resolver) flip() string {
	v, ok := prng.Pick(r.rng, "flip", r.options.flip(), strIdentity)
	if !ok {
		v = "none"
	}
	r.record("flip", v)
	return v
}

func (r *resolver) fontFamily() string {
	v, ok := prng.Pick(r.rng, "fontFamily", r.options.fontFamily(), strIdentity)
	if !ok {
		v = "system-ui"
	}
	r.record("fontFamily", v)
	return v
}

func (r *resolver) fontWeight() float64 {
	v, ok := prng.Pick(r.rng, "fontWeight", r.options.fontWeight(), prng.NumString)
	if !ok {
		v = 400
	}
	r.record("fontWeight", v)
	return v
}

func (r *resolver) scale() float64 { return r.floatOpt("scale", r.options.scale(), 1) }
func (r *resolver) borderRadius() float64 {
	return r.floatOpt("borderRadius", r.options.borderRadius(), 0)
}
func (r *resolver) rotate() float64     { return r.floatOpt("rotate", r.options.rotate(), 0) }
func (r *resolver) translateX() float64 { return r.floatOpt("translateX", r.options.translateX(), 0) }
func (r *resolver) translateY() float64 { return r.floatOpt("translateY", r.options.translateY(), 0) }

// variant selects a variant for the given component, or ("", false) when the
// component is unknown or rolled invisible.
func (r *resolver) variant(name string) (string, bool) {
	v, ok := r.resolveVariant(name)
	if ok {
		r.record(name+"Variant", v)
	} else {
		r.record(name+"Variant", nil)
	}
	return v, ok
}

func (r *resolver) resolveVariant(name string) (string, bool) {
	comp, ok := r.style.Components()[name]
	if !ok {
		return "", false
	}
	if !r.isVisible(name, comp) {
		return "", false
	}

	return r.rng.WeightedPick(name+"Variant", r.variantWeights(comp))
}

// variantWeights builds the name → weight map the PRNG draws a variant from. The
// per-component ${name}Variant option is more specific than the global tags
// filter, so it takes precedence: when set, it fully governs the component's pool
// (its named variants, weighted by the option) and the tags filter is ignored for
// that component. The tags filter applies only where the user gave no explicit
// ${name}Variant, and falls back to every variant when neither is set. Names the
// style does not define are dropped, and an empty ${name}Variant (or an empty tag
// result) yields no variant.
func (r *resolver) variantWeights(comp *style.Component) map[string]float64 {
	variants := comp.Variants()
	named, hasNamed := r.options.componentVariant(comp.SourceName())
	weights := map[string]float64{}

	if hasNamed {
		for v, w := range named {
			if _, exists := variants[v]; exists {
				weights[v] = w
			}
		}
		return weights
	}

	if len(r.options.tags()) > 0 {
		for _, v := range r.tagFilteredNames(variants) {
			weights[v] = variants[v].WeightOr1()
		}
		return weights
	}

	for v, variant := range variants {
		weights[v] = variant.WeightOr1()
	}
	return weights
}

type tagAllow struct {
	category string
	values   []string
}

type tagDisallow struct {
	category string
	value    string
	hasValue bool
}

// tagFilter holds the parsed options.tags tokens grouped by role. bareDisallows
// is the subset of disallows carrying no value, kept as a lookup set for the
// per-component narrowing.
type tagFilter struct {
	allowGroups   []*tagAllow
	bares         []string
	disallows     []tagDisallow
	bareDisallows map[string]bool
}

// tagFilter classifies the parsed options.tags tokens into the allow groups,
// bare requirements and disallows the filter is composed from. The result
// depends only on the options, never on a component, so it is computed once per
// avatar rather than rebuilt for each of a style's components.
func (r *resolver) tagFilter() *tagFilter {
	if r.tags != nil {
		return r.tags
	}

	f := &tagFilter{bareDisallows: map[string]bool{}}
	allowIndex := map[string]*tagAllow{}
	bareSeen := map[string]bool{}

	for _, tok := range r.options.tags() {
		if tok.negated {
			f.disallows = append(f.disallows, tagDisallow{category: tok.category, value: tok.value, hasValue: tok.hasValue})
			if !tok.hasValue {
				f.bareDisallows[tok.category] = true
			}
		} else if tok.hasValue {
			grp, ok := allowIndex[tok.category]
			if !ok {
				grp = &tagAllow{category: tok.category}
				allowIndex[tok.category] = grp
				f.allowGroups = append(f.allowGroups, grp)
			}
			grp.values = append(grp.values, tok.value)
		} else if !bareSeen[tok.category] {
			bareSeen[tok.category] = true
			f.bares = append(f.bares, tok.category)
		}
	}

	r.tags = f

	return f
}

// tagFilteredNames narrows a component's variants to the names satisfying the
// global tags filter, applying the parsed options.tags tokens in one pass over
// the pool:
//
//   - A positive cat:value token is an axis-scoped allow. Within each category
//     some allow mentions, a variant is kept only if it carries no tag in that
//     category (untouched) or matches one of the allowed values (OR within the
//     category). Distinct allowed categories combine with AND, and a category no
//     allow mentions is left unconstrained.
//   - A bare positive cat token requires the category: it drops variants that
//     carry no tag in cat. It only binds where the category is in use — a
//     component with no cat tag on any variant is left untouched, so animation
//     turns on a style's opt-in animation without erasing the components that
//     know nothing about it.
//   - A negative !cat / !cat:value token disallows, dropping every variant
//     carrying any tag in cat (bare) or the exact cat:value tag. Disallows are
//     checked alongside allows but always win.
//
// The result order is irrelevant: WeightedPick sorts the names internally.
func (r *resolver) tagFilteredNames(variants map[string]style.ComponentVariant) []string {
	f := r.tagFilter()
	allowGroups, bares, disallows := f.allowGroups, f.bares, f.disallows

	// A bare token only binds where its category is in use, so this narrowing —
	// unlike the classification — is genuinely per-component.
	var required []string
	for _, category := range bares {
		if f.bareDisallows[category] {
			continue
		}
		for _, variant := range variants {
			if variant.HasTag(category, "", false) {
				required = append(required, category)
				break
			}
		}
	}

	var names []string
	for name, variant := range variants {
		allowed := true
		for _, category := range required {
			if !variant.HasTag(category, "", false) {
				allowed = false
				break
			}
		}
		if !allowed {
			continue
		}

		for _, grp := range allowGroups {
			if !variant.HasTag(grp.category, "", false) {
				continue
			}
			matched := false
			for _, value := range grp.values {
				if variant.HasTag(grp.category, value, true) {
					matched = true
					break
				}
			}
			if !matched {
				allowed = false
				break
			}
		}
		if !allowed {
			continue
		}

		disallowed := false
		for _, ex := range disallows {
			if variant.HasTag(ex.category, ex.value, ex.hasValue) {
				disallowed = true
				break
			}
		}

		if !disallowed {
			names = append(names, name)
		}
	}

	return names
}

func (r *resolver) color(name string) ([]string, error) {
	key := name + "Color"

	// Memoize like the JS #memo: a color already resolved this pass is returned
	// from the snapshot instead of being recomputed. Without it, a graph where
	// each color references the next via both contrastTo and notEqualTo
	// re-resolves exponentially (a schema-valid DoS).
	if cached, ok := r.cachedColor(key); ok {
		return cached, nil
	}

	value, err := r.resolveColor(name)
	if err != nil {
		return nil, err
	}
	r.record(key, value)
	return value, nil
}

func (r *resolver) cachedColor(key string) ([]string, bool) {
	v, ok := r.result[key]
	if !ok {
		return nil, false
	}
	if arr, ok := v.([]string); ok {
		return arr, true
	}
	return nil, true
}

func (r *resolver) colorFill(name string) string {
	v, ok := prng.Pick(r.rng, name+"ColorFill", r.options.colorFill(name), strIdentity)
	if !ok {
		v = "solid"
	}
	r.record(name+"ColorFill", v)
	return v
}

func (r *resolver) colorAngle(name string) float64 {
	return r.floatOpt(name+"ColorAngle", r.options.colorAngle(name), 0)
}

// colorOrder returns the user's ${name}ColorOrder, or "random" when unset.
// Deliberately not recorded: unlike colorFill this is no PRNG pick, so it
// stays out of the resolved snapshot.
func (r *resolver) colorOrder(name string) string {
	if v := r.options.colorOrder(name); v != "" {
		return v
	}
	return style.ColorOrderRandom
}

// componentTransform returns (rotate, translateX, translateY, scale) for a
// component, each recorded as ${name}Rotate / ${name}TranslateX / … in the
// snapshot.
func (r *resolver) componentTransform(name string) (rotate, translateX, translateY, scale float64) {
	comp := r.style.Components()[name]

	var rRange, sRange, txRange, tyRange *prng.Range
	if comp != nil {
		rRange = comp.Rotate()
		sRange = comp.Scale()
		txRange = comp.TranslateX()
		tyRange = comp.TranslateY()
	}

	rotate = r.floatOpt(name+"Rotate", rRange, 0)
	translateX = r.floatOpt(name+"TranslateX", txRange, 0)
	translateY = r.floatOpt(name+"TranslateY", tyRange, 0)
	scale = r.floatOpt(name+"Scale", sRange, 1)
	return
}

// resolved returns an informational snapshot of every value the resolver
// picked. Unset (nil) entries are filtered out; the raw seed is excluded.
func (r *resolver) resolved() map[string]any {
	out := make(map[string]any, len(r.result))
	for k, v := range r.result {
		if v != nil {
			out[k] = v
		}
	}
	return out
}

// resolvedOrder returns the snapshot keys with non-nil values in the order they
// were first recorded, matching the JS/Rust resolution order for the JSON
// envelope.
func (r *resolver) resolvedOrder() []string {
	out := make([]string, 0, len(r.order))
	for _, k := range r.order {
		if r.result[k] != nil {
			out = append(out, k)
		}
	}
	return out
}

func (r *resolver) probability(comp *style.Component) float64 {
	if p := r.options.componentProbability(comp.SourceName()); p != nil {
		return *p
	}
	return comp.Probability()
}

func (r *resolver) isVisible(name string, comp *style.Component) bool {
	return r.rng.Bool(name+"Probability", r.probability(comp))
}

// resolveColor resolves a named color to its final stop list, applying
// contrast sorting and notEqualTo filtering from the style definition.
// Circular references between colors (e.g. a.contrastTo = b, b.contrastTo = a)
// are detected via the colorResolving stack and reported as a
// CircularColorReferenceError.
//
// A user-set ${name}ColorOrder: "fixed" pins the candidates to their given
// order, whether they come from the ${name}Color option or from the style's
// palette: the shuffle and the contrast sort are skipped (notEqualTo filtering
// still applies), and the gradient stop count defaults to the number of
// candidates instead of 2.
func (r *resolver) resolveColor(name string) ([]string, error) {
	styleColor, hasStyleColor := r.style.Colors()[name]

	userColors, hasUserColors := r.options.color(name)
	var source []string
	if hasUserColors {
		source = userColors
	} else if hasStyleColor {
		source = styleColor.Values
	}

	candidates := make([]string, len(source))
	for i, c := range source {
		candidates[i] = color.ToHex(c)
	}

	fixed := r.colorOrder(name) == style.ColorOrderFixed
	fill := r.colorFill(name)
	stops := 1
	if fill != "solid" {
		fallback := 2
		if fixed {
			fallback = len(candidates)
		}
		stops = r.colorFillStops(name, fallback)
	}

	if !hasStyleColor {
		return takeN(r.orderColors(name, candidates, fixed), stops), nil
	}

	// Detect circular references (e.g. a.contrastTo = b, b.contrastTo = a).
	for _, n := range r.colorResolving {
		if n == name {
			chain := append(append([]string{}, r.colorResolving...), name)
			return nil, &errs.CircularColorReferenceError{Chain: chain}
		}
	}

	r.colorResolving = append(r.colorResolving, name)
	err := r.applyColorConstraints(styleColor, &candidates, fixed)
	r.colorResolving = r.colorResolving[:len(r.colorResolving)-1]
	if err != nil {
		return nil, err
	}

	// Skip the shuffle when sorted by contrast, to preserve that ordering.
	ordered := candidates
	if styleColor.ContrastTo == "" {
		ordered = r.orderColors(name, candidates, fixed)
	}

	return takeN(ordered, stops), nil
}

// orderColors applies ${name}ColorOrder to the candidate list. "random"
// shuffles via the PRNG, "fixed" keeps the candidates as given, duplicates
// included.
func (r *resolver) orderColors(name string, candidates []string, fixed bool) []string {
	if fixed {
		return candidates
	}
	return r.rng.Shuffle(name+"Color", candidates)
}

func (r *resolver) applyColorConstraints(styleColor style.ColorDef, candidates *[]string, fixed bool) error {
	if styleColor.ContrastTo != "" && !fixed {
		refColors, err := r.color(styleColor.ContrastTo)
		if err != nil {
			return err
		}
		if len(refColors) > 0 {
			*candidates = color.SortByContrast(*candidates, refColors[0])
		}
	}

	if len(styleColor.NotEqualTo) > 0 {
		var excluded []string
		for _, ref := range styleColor.NotEqualTo {
			cols, err := r.color(ref)
			if err != nil {
				return err
			}
			excluded = append(excluded, cols...)
		}
		*candidates = color.FilterNotEqualTo(*candidates, excluded)
	}

	return nil
}

func (r *resolver) colorFillStops(name string, fallback int) int {
	if rng := r.options.colorFillStops(name); rng != nil {
		n := r.rng.Integer(name+"ColorFillStops", rng)
		if n < 0 {
			n = 0
		}
		return n
	}
	return fallback
}

func (r *resolver) floatOpt(key string, rng *prng.Range, fallback float64) float64 {
	value := fallback
	if rng != nil {
		value = r.rng.Float(key, rng)
	}
	r.record(key, value)
	return value
}

// takeN returns a copy of the first n elements of s (or all of them if n is
// larger), mirroring the JS slice(0, stops).
func takeN(s []string, n int) []string {
	if n > len(s) {
		n = len(s)
	}
	if n < 0 {
		n = 0
	}
	out := make([]string, n)
	copy(out, s[:n])
	return out
}
